Let the mixer configure the channels lineout modes (#1590, #1594)

kcgen opened the pull request:
This adds the ability to adjust a given channel's lineout mode via the mixer's command-line.

Similar to how the mixer already provides volume control for the channels, this adds the following line-out modes (per emulated audio device/channel):

 - `stereo`: normal 1:1 connection (left-to-left, right-to-right). This is the default.
 - `reverse`: the channel's left line is fed into the mixer's right line (and vice-versa).
 - `leftmono`: both of the channel's lines are fed into mixer's left line.
 - `rightmono`: both of the channel's lines are fed into mixer's right line.

This avoids having to further pollute the conf file with reverse stereo settings in each audio device section (ie: "mt32_reverse", "sb_reverse", and so on).

Likewise, real hardware such as CD players, headphones, and TVs similarly don't include a toggle switch to reverse their stereo output. Instead, people who care about this stuff adjust the physical RCA lineouts feeding into their amplifier (or logically in the amplifier itself).

**Added to `mixer /?`:**

![2022-03-22_23-06](https://user-images.githubusercontent.com/1557255/159640690-28caddc3-40cf-456a-a36d-5d38cbf3eeef.png)

**Examples:**

![2022-03-22_23-07](https://user-images.githubusercontent.com/1557255/159640721-2c006d8c-9d05-432c-9546-20477f522a95.png)

**Issues:**
 - Fixes #1590
 - Fixes #1594

Wengier reviewed (approved):
Both the code itself, help messages, and the function look good. Nice work!

shermp commented on src/hardware/mixer.cpp:
```diff
@@ -651,6 +655,35 @@ void MixerChannel::FillUp()
+		return right_lineout ? "RightMono" : "LeftMono";
+}
+
+bool MixerChannel::SetLineoutMode(std::string lineout_mode)
+{
+	lowcase(lineout_mode);
+	if (lineout_mode == "stereo") {
+		left_lineout = STEREO_LEFT_INDEX;
+		right_lineout = STEREO_RIGHT_INDEX;
+	} else if (lineout_mode == "reverse") {
+		left_lineout = STEREO_RIGHT_INDEX;
+		right_lineout = STEREO_LEFT_INDEX;
+	} else if (lineout_mode == "leftmono") {
+		left_lineout = STEREO_LEFT_INDEX;
+		right_lineout = STEREO_LEFT_INDEX;
+	} else if (lineout_mode == "rightmono") {
+		left_lineout = STEREO_RIGHT_INDEX;
+		right_lineout = STEREO_RIGHT_INDEX;
+	} else {
+		return false;
+	}
+	return true;
+}
```
How about another private member function like `SetLineouts(uint8_t left, uint8_t right)` to set the lineout variables? Not a biggie, but would reduce duplication a bit.

(or `SetLineouts(uint8_t right, uint8_t left)` if you're an audiologist, because audiologists are weird like that, judging by my last audiogram)

kcgen replied:
Good news is the enum type allowed for the creation of constexpr structs for each of the lineout modes:

![2022-03-24_17-13](https://user-images.githubusercontent.com/1557255/160030150-cc5cf5e2-c27a-416b-9258-768abc291cef.png)

Which Allowed for one-time (and exact) assignment:

![2022-03-24_17-14](https://user-images.githubusercontent.com/1557255/160030217-48d1cf80-533c-4b9b-8a9d-319c06750493.png)

kcgen commented on include/mixer.h:
```diff
@@ -61,6 +61,9 @@ extern Bit8u MixTemp[MIXER_BUFSIZE];
 
+constexpr uint8_t STEREO_LEFT_INDEX = 0;
+constexpr uint8_t STEREO_RIGHT_INDEX = 1;
+
```
I definitely agree.

It's now moved into a `LINE_INDEX` enum, and this is now used in the MSCDEX channel map call as well.  Even though we're dealing with a one and a zero, it's the principle that I think generally applies, regardless how many or how few variations wer dealing with. 

Thanks for the suggestion @shermp :+1:

kcgen commented:
@shermp , I think I've addressed your feedback. 

If anyone plans to help more with this - be it testing, code review, or general feedback - please let me know!

If not, will plan to merge in the next day or so.

johnnovak commented:
It seems you haven't added the description of this new option to the help text of the `MIXER` command, was that intentional?

kcgen commented:
Oh, darn -- looks like I lost some commits; maybe I stashed them and forgot to pop them back. 
Atleast I took a screenshot of the help screen in the PR description ; will recreate from that. 
I remember also using a switch in a later commit too, instead of the if/else worm.
Will re-add; thanks, @johnnovak !

johnnovak commented:
@kcgen just a question about the intention behind `leftmono` and `rightmono`.

The ability to reverse the stereo channels with `reverse` makes perfect sense, it's a handy feature. I understand what `leftmono` and `rightmono` is technically doing, but I struggle to think of any real-world use-case for them. Can you name a few please?

The reason I'm asking is because it could interfere with the crossfeed feature I'm working on in weird and unexpected ways. E.g. a number of sound cards produce mono output (SB1 & SB2, all "speaker" types, and Innovation as well). These currently show up in the mixer as "Stereo", which is not quite correct in my opinion. Then you can set linemode to `leftmono` and `rightmono` on them which surely works as internally we're only dealing with stereo streams. But it begs the questions, why would anyone want to do that in the first place?

My thinking was to "tag" channels with "features", e.g. for channels that we know for certain can only produce mono output, we tag them with `ChannelFeature::Mono` at creation time or something, then they will show up in the mixer as "Mono". For mono channels it makes no sense to apply crossfeed of reverse stereo linemode. This would be an easy and simple to understand concept (crossfeed and reverse stereo only works for "Stereo" channels), but the `leftmono` and `rightmono` options throw a spanner in the works, they kinda confuse everything (should we allow `leftmono`/`rightmono` for "real" stereo streams only? or to mono streams as well? what does it even mean to set a mono channel to `leftmono`? is it even desirable? etc.)

So yeah, maybe I'm missing some functionality here that I never needed, but to make only allowing `stereo` and `reverse` linesmodes make sense, and then we could tag the channels as "mono" or "stereo" and everything would be straightforward to the user.

kcgen commented:
For people with a single desktop speaker plugged into the audio-out jack, the RCA mini-plug will only get the signal from either the left or right channel. So this setting lets these users send both the left and right audio channels into a single "side" - so they actually hear the full composition (albiet, in mono).

 - It's definitely a corner case, and very likely unused
 - Desktop OS mixers are now sufficiently advanced on all supported OSes that they can do this for us

Cross-feed is 100% applicable and will benefit everyone - so please feel free to drop the mono feature ! (if anyone really really needs it, we can help then use their desktop mixer).

johnnovak commented:
> For people who plug a single desktop speaker into their audio-out jack, the RCA mini-plug will only get the signal from either the left or right channel. So this setting lets these users send both audio channels into whatever "side" their speaker's plugged into - so they actually hear the full composition (albeit, in mono).
> 
>     * It's definitely a corner case, and very likely unused
> 
>     * Desktop OS mixers are now sufficiently advanced on all supported OSes that they can do this for us
> 
> 
> Cross-feed is 100% applicable and will benefit everyone - so please feel free to drop the mono modes ! (if anyone really really needs it, we can help then use their desktop mixer 🍻 ).

Thanks for the explanation @kcgen, I agree we can safely drop the mono modes as it's a very rare edge case. That would simplify things a lot!
